modules/apps/static/portal/portal-db-partition-impl/src/main/java/com/liferay/portal/db/partition/internal/DBPartitionHelperImpl.java

```diff
@@ -0,0 +1,212 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portal.db.partition.internal;
+
+import com.liferay.petra.string.StringBundler;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.dao.db.DB;
+import com.liferay.portal.kernel.dao.db.DBInspector;
+import com.liferay.portal.kernel.dao.db.DBManagerUtil;
+import com.liferay.portal.kernel.dao.db.DBType;
+import com.liferay.portal.kernel.dao.db.partition.DBPartitionHelper;
+import com.liferay.portal.kernel.dao.jdbc.CurrentConnectionUtil;
+import com.liferay.portal.kernel.dao.orm.ORMException;
+import com.liferay.portal.kernel.model.CompanyConstants;
+import com.liferay.portal.kernel.security.auth.CompanyThreadLocal;
+import com.liferay.portal.kernel.util.GetterUtil;
+import com.liferay.portal.kernel.util.InfrastructureUtil;
+import com.liferay.portal.kernel.util.PropsUtil;
+
+import java.sql.Connection;
+import java.sql.DatabaseMetaData;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+
+/**
+ * @author Alberto Chaparro
+ */
+@Component(immediate = true, service = {})
+public class DBPartitionHelperImpl implements DBPartitionHelper {
+
+	@Override
+	public boolean addPartition(long companyId) {
+		if (companyId == _defaultCompanyId) {
+			return false;
+		}
+
+		Connection connection = CurrentConnectionUtil.getConnection(
+			InfrastructureUtil.getDataSource());
+
+		String schemaName = "company" + companyId;
```

Comment:
@achaparro we should make this generated schema prefix configurable too, default to "company" is fine. Just we should allow user to change it.

Reply:
This is only for us, for our Saas implementation so I assumed we can't allow users to change it. The problem of making it configurable is that, if they change it, we need to stop the server and rename the database schema. Also, if we allow modifying it we need to keep records about the different schema names. Having a hardcoded name simplifies the solution and I guess we could do it since the schema name is not relevant for the user.

Reply:
When I say user, I actually mean our internal user. The problem for hardcode is really about sharing the DBMS. The same DBMS may be shared by different portal setup, which will require multiple-groups of "schemas". If they all just named as "company" + a number, we won't be able to tell which schema belongs to which group just by looking at the name.

Reply:
I understand that part. We can make configurable the prefix (but we should warn that any change in that property requires to shut down the portal rename the schemas first) so we would have something like this:
{INSTANCE_PREFFIX}+CompanyId

What do you think? Any suggestion?

Reply:
That's fine. I doubt it is needed to warn people this, just state the purpose of this property in the comment should be enough. Renaming it has the same affect as changing schema name in your jdbc url, it is obviously that you can not assume portal would know how to rename the schema automatically.